src/libratbag.c

```diff
@@ -1909,3 +1909,132 @@ ratbag_button_macro_new(const char *name)
+	struct ratbag_modifier_mapping *mapping;
+	int i;
+
+	macro = ratbag_button_macro_new("key");
+	i = 0;
+
+	ARRAY_FOR_EACH(ratbag_modifier_mapping, mapping) {
+		if (modifiers & mapping->modifier_mask)
+			ratbag_button_macro_set_event(macro,
+						      i++,
+						      RATBAG_MACRO_EVENT_KEY_PRESSED,
+						      mapping->key);
+	}
+
+	ratbag_button_macro_set_event(macro,
+				      i++,
+				      RATBAG_MACRO_EVENT_KEY_PRESSED,
+				      key);
+	ratbag_button_macro_set_event(macro,
+				      i++,
+				      RATBAG_MACRO_EVENT_KEY_RELEASED,
+				      key);
+
+	ARRAY_FOR_EACH(ratbag_modifier_mapping, mapping) {
+		if (modifiers & mapping->modifier_mask)
+			ratbag_button_macro_set_event(macro,
+						      i++,
+						      RATBAG_MACRO_EVENT_KEY_RELEASED,
+						      mapping->key);
+	}
+
+	ratbag_button_copy_macro(button, macro);
+	ratbag_button_macro_unref(macro);
+
+	return 0;
+}
+
+int
+ratbag_action_keycode_from_macro(struct ratbag_button_action *action,
+				 unsigned int *key_out,
+				 unsigned int *modifiers_out)
+{
+	struct ratbag_macro *macro = action->macro;
+	unsigned int key = KEY_RESERVED;
+	unsigned int modifiers = 0;
+	unsigned int i;
+
+	if (!macro || action->type != RATBAG_BUTTON_ACTION_TYPE_MACRO)
+		return -EINVAL;
+
+	if (macro->events[0].type == RATBAG_MACRO_EVENT_NONE)
+		return -EINVAL;
+
+	for (i = 0; i < MAX_MACRO_EVENTS; i++) {
+		struct ratbag_macro_event event;
+
+		event = macro->events[i];
+		switch (event.type) {
+		case RATBAG_MACRO_EVENT_INVALID:
+			return -EINVAL;
+		case RATBAG_MACRO_EVENT_NONE:
+			return 0;
+		case RATBAG_MACRO_EVENT_KEY_PRESSED:
+			switch(event.event.key) {
+			case KEY_LEFTCTRL: modifiers |= MODIFIER_LEFTCTRL; break;
+			case KEY_LEFTSHIFT: modifiers |= MODIFIER_LEFTSHIFT; break;
+			case KEY_LEFTALT: modifiers |= MODIFIER_LEFTALT; break;
+			case KEY_LEFTMETA: modifiers |= MODIFIER_LEFTMETA; break;
+			case KEY_RIGHTCTRL: modifiers |= MODIFIER_RIGHTCTRL; break;
+			case KEY_RIGHTSHIFT: modifiers |= MODIFIER_RIGHTSHIFT; break;
+			case KEY_RIGHTALT: modifiers |= MODIFIER_RIGHTALT; break;
+			case KEY_RIGHTMETA: modifiers |= MODIFIER_RIGHTMETA; break;
+			default:
+				if (key != KEY_RESERVED)
+					return -EINVAL;
+
+				key = event.event.key;
+			}
+			break;
+		case RATBAG_MACRO_EVENT_KEY_RELEASED:
+			switch(event.event.key) {
+			case KEY_LEFTCTRL: modifiers &= ~MODIFIER_LEFTCTRL; break;
+			case KEY_LEFTSHIFT: modifiers &= ~MODIFIER_LEFTSHIFT; break;
+			case KEY_LEFTALT: modifiers &= ~MODIFIER_LEFTALT; break;
+			case KEY_LEFTMETA: modifiers &= ~MODIFIER_LEFTMETA; break;
+			case KEY_RIGHTCTRL: modifiers &= ~MODIFIER_RIGHTCTRL; break;
+			case KEY_RIGHTSHIFT: modifiers &= ~MODIFIER_RIGHTSHIFT; break;
+			case KEY_RIGHTALT: modifiers &= ~MODIFIER_RIGHTALT; break;
+			case KEY_RIGHTMETA: modifiers &= ~MODIFIER_RIGHTMETA; break;
+			default:
+				if (event.event.key != key)
+					return -EINVAL;
+
+				*key_out = key;
+				*modifiers_out = modifiers;
+				return 1;
```

Comment:
Shouldn't you check before returning if there are no other key pressed?
I can imagine a macro with `Ctrl-C Ctrl-V` that would be translated into `Ctrl-V` this way.

Reply:
The function is only intended to be used when we use a macro to do a simple keybinding. We should have caps (#271) that the client should respect so we do not try to set a `Ctrl-C Ctrl-V` on a device which can only do one key. I added a new guard to `ratbag_action_keycode_from_macro` which checks that the macro only contain one key.